lib/std/uniquelabels/unique_labels.go

```diff
@@ -0,0 +1,209 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package uniquelabels
+
+import (
+	"encoding/json"
+	"fmt"
+	"iter"
+	"maps"
+
+	"github.com/projectcalico/calico/lib/std/uniquestr"
+)
+
+type handleMap = map[uniquestr.Handle]uniquestr.Handle
+
+var (
+	Nil   = Map{m: nil}
+	Empty = Map{m: map[uniquestr.Handle]uniquestr.Handle{}}
+)
+
+// Map is a *read only* string-to-string map that interns keys and values.
+// When marshalled as JSON, uses the same representation as a normal
+// map[string]string.  When unmarshalled, keys and values get interned/deduped.
+//
+// To allow drop-in replacement of a map[string]string, Map has unique
+// representations of nil (Nil) and empty (Empty). The zero value of a Map is
+// the Nil value; it can be detected with IsNil(). Map is a value type, which
+// precludes using Go's normal nil.  Making Map a pointer type would increase
+// overheads; making map an alias for map[uniquestr.Handle]uniquestr.Handle
+// would break the read-only property, preventing future optimisations.
+//
+// Uses uniquestr.Handle internally, so it is most efficient to query the map
+// using AllHandles() and GetHandle().
+type Map struct {
+	_ [0]func() // Explicitly non-comparable; must use Equals() method.
+	m handleMap
+}
+
+// Make makes an interned copy of the given map.  In order to benefit from
+// interning the map, the original map must be discarded and only the interned
+// copy should be kept.
+//
+// If passed nil, returns the zero value of Map.  If passed an empty map,
+// returns the singleton Empty.
+func Make(m map[string]string) Map {
+	var hm handleMap
+	if m == nil {
+		return Nil
+	}
+	if len(m) == 0 {
+		return Empty
+	}
+	hm = make(handleMap, len(m))
+	for k, v := range m {
+		hm[uniquestr.Make(k)] = uniquestr.Make(v)
+	}
+	return Map{m: hm}
+}
+
+// Equals returns true if the map contains the same key/value pairs as the
+// other map.  In line with maps.Equal, Nil and Empty compare equal.
+func (i Map) Equals(other Map) bool {
+	return maps.Equal(i.m, other.m)
+}
+
+// MarshalJSON implements the json.Marshaler interface. Must be defined on the
+// value receiver so that Map can be embedded in other structs.
+func (i Map) MarshalJSON() ([]byte, error) {
+	return json.Marshal(i.m)
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface.  Must be defined on
+// the pointer receiver so that it can have side effects.
+func (i *Map) UnmarshalJSON(data []byte) error {
+	var hm handleMap
+	if err := json.Unmarshal(data, &hm); err != nil {
+		return err
+	}
+	i.m = hm
+	return nil
+}
+
+func (i Map) AllHandles() iter.Seq2[uniquestr.Handle, uniquestr.Handle] {
+	return func(yield func(uniquestr.Handle, uniquestr.Handle) bool) {
+		for k, v := range i.m {
+			if !yield(uniquestr.Handle(k), uniquestr.Handle(v)) {
```

Comment:
I'm not entirely sure I follow why we need to do a type conversion of `k` and `v` to `uniquestr.Handle` here - aren't these already coming from `i.m`, which is itself a `map[uniquestr.Handle]uniquestr.Handle`?

Reply:
Yep, redundant now, came from a a refactoring

Reply:
Fixed